Add more options for the WooCommerce coupon block in the email editor - MAILPOET-4762

triple0t opened the pull request:
## Description

Add coupon settings to the Email editor WooCommerce coupon block.

## Code review notes

Please start reviewing from https://github.com/mailpoet/mailpoet/commit/92b9d11fe952abd76f8d90d780464a9325948f36

## QA notes

You need to enable Coupon block experimental features here: `/wp-admin/admin.php?page=mailpoet-experimental`

* On the email editor
* Select the coupon block
* Click on the Usage restriction or Usage limits
* Add some options
* Send the newsletter
* Your newsletter should have the new coupon code. 
* You may verify the coupon options at `/wp-admin/edit.php?post_type=shop_coupon`, select the coupon code found in the newsletter

## Linked PRs

Dependent on https://github.com/mailpoet/mailpoet/pull/4636

## Linked tickets

[MAILPOET-4762](https://mailpoet.atlassian.net/browse/MAILPOET-4762)




[MAILPOET-4762]: https://mailpoet.atlassian.net/browse/MAILPOET-4762?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

triple0t committed:
Add other WooCommerce coupon option methods

MAILPOET-4762

triple0t committed:
Add basic accordion to group coupon data and free shipping field

MAILPOET-4762

triple0t committed:
Add Usage limit fields and half of Usage restriction fields

MAILPOET-4762

triple0t committed:
Add Product categories and Exclude categories Usage restriction fields

MAILPOET-4762

triple0t committed:
Add Products and Exclude products Usage restriction fields

MAILPOET-4762

triple0t committed:
Fix tests

MAILPOET-4762

triple0t committed:
Add validation for some fields

MAILPOET-4762

rodrigoprimo commented on mailpoet/views/newsletter/templates/blocks/coupon/settings.hbs:
```diff
@@ -59,6 +59,7 @@
           <div class="mailpoet_form_field_title mailpoet_form_field_title_inline"><%= __('Minimum spend') %></div>
           <div class="mailpoet_form_field_input_option">
             <input type="text" name="minimum-amount" class="mailpoet_input mailpoet_field_coupon_minimum_amount" value="{{ model.minimumAmount }}" placeholder="<%= __('No minimum')%>" />
+            <span class="mailpoet_error mailpoet_hidden"> <%= __('Please enter with one monetary decimal point (.) without thousand separators and currency symbols.') %> </span>
```
If I remember correctly, WooCommerce supports different decimal separators and not just `.`. Maybe we need to account for that in this error message and display whatever the user has set as their decimal separator?

triple0t replied:
Noted

rodrigoprimo commented on mailpoet/views/newsletter/templates/blocks/coupon/settings.hbs:
```diff
@@ -68,6 +69,7 @@
           <div class="mailpoet_form_field_title mailpoet_form_field_title_inline"><%= __('Maximum spend') %></div>
           <div class="mailpoet_form_field_input_option">
             <input type="text" name="maximum-amount" class="mailpoet_input mailpoet_field_coupon_maximum_amount" value="{{ model.maximumAmount }}" placeholder="<%= __('No Maximum')%>" />
+            <span class="mailpoet_error mailpoet_hidden"> <%= __('Please enter with one monetary decimal point (.) without thousand separators and currency symbols.') %> </span>
```
Same as mentioned below

triple0t replied:
Noted

rodrigoprimo commented on mailpoet/views/newsletter/templates/blocks/coupon/settings.hbs:
```diff
@@ -142,6 +144,7 @@
           <div class="mailpoet_form_field_title mailpoet_form_field_title_inline"><%= __('Allowed emails') %></div>
           <div class="mailpoet_form_field_input_option">
             <input type="email" name="email-restrictions" class="mailpoet_input mailpoet_field_coupon_email_restrictions" value="{{ model.emailRestrictions }}" placeholder="<%= __('No restrictions')%>" multiple="multiple" />
+            <span class="mailpoet_error mailpoet_hidden"> </span>
```
Is this <span> needed for something that I'm not seeing?

triple0t replied:
Yeah. This is used to display the error message

rodrigoprimo commented on mailpoet/views/newsletter/templates/blocks/coupon/settings.hbs:
```diff
@@ -54,7 +54,53 @@
+        <label>
+          <div class="mailpoet_form_field_title mailpoet_form_field_title_inline"><%= __('Minimum spend') %></div>
+          <div class="mailpoet_form_field_input_option">
+            <input type="text" name="minimum-amount" class="mailpoet_input mailpoet_field_coupon_minimum_amount" value="{{ model.minimumAmount }}" placeholder="<%= __('No minimum')%>" />
```
I'm not super familiar with the technical stack of the email editor, so I'm unsure about how sanitizing input and escaping output works here. Do we need to do something about it or is one of the libraries that we are using already taking care of both?

triple0t replied:
> Do we need to do something about it or is one of the libraries that we are using already taking care of both?

Yes, we have already taken care of this.

But a few improvements can be made, and I will implement that in the next commit.

rodrigoprimo replied:
Thanks, could you please detail what we are already doing to take care of this and what few improvements still needed to be made that you implemented? This information will help when reviewing the code.

Also, I wonder if there is a reason to combine the changes to improve escaping and sanitizing with the changes to support different decimal separators, and I believe indentation changes as well, in a single commit (cabdbcd)? From my perspective, this makes reviewing this PR a bit harder and it might make understanding the changes proposed here slightly more difficult if in the future we need to debug this part of the code. I would suggest splitting those changes into multiple commits. What do you think?

triple0t replied:
> could you please detail what we are already doing to take care of this

We perform the sanitizing (and user input escape) when the Newsletter is saved [here](https://github.com/mailpoet/mailpoet/blob/trunk/mailpoet/lib/Newsletter/NewsletterSaveController.php#L116).

We are using Handlebarsjs for our templates and escaping the output with `{{}}`. [read more here](https://handlebarsjs.com/guide/expressions.html#html-escaping)

> what few improvements still needed to be made that you implemented?

There's generally no need to add an escape here as it serves for Html placeholder not user input.
I thought to add an [improvement](https://github.com/mailpoet/mailpoet/blob/cabdbcd4b9b495a24874303b2cd059ca32a173b7/mailpoet/views/newsletter/templates/blocks/coupon/settings.hbs#L113) there to be safe.

triple0t replied:
> Also, I wonder if there is a reason to combine the changes to improve escaping and sanitizing with the changes to support different decimal separators, and I believe indentation changes as well, in a single commit (https://github.com/mailpoet/mailpoet/commit/cabdbcd4b9b495a24874303b2cd059ca32a173b7)? From my perspective, this makes reviewing this PR a bit harder and it might make understanding the changes proposed here slightly more difficult if in the future we need to debug this part of the code. I would suggest splitting those changes into multiple commits. What do you think?

Done

rodrigoprimo reviewed (commented):
Good job, @triple0t. I left a few questions, and there is at least one minor thing that I believe we should address before moving this PR to QA.

triple0t committed:
Use WooCommerce decimal separators when validating input fields

We need to use the decimal separators selected by the user for our validation. We are also using the validation used on the adding/editing coupon page

MAILPOET-4762

triple0t committed:
Escape the content for placeholder

MAILPOET-4762

triple0t committed:
Update tests and remove unused variables

MAILPOET-4762

triple0t committed:
Stop saving invalid input values

MAILPOET-4762

rodrigoprimo commented:
Thanks for the update, @triple0t. I could confirm that the error string is now using whatever decimal separator the user configured in WooCommerce.

I pinged you on Slack for us to discuss a final detail before moving this forward.

Also, now that the other PR was merged, you might want to add to the QA instructions of this PR that it is needed to enable its feature flag to be able to test the coupon block.

rodrigoprimo reviewed (dismissed):
@triple0t, I'm approving this PR but leaving it assigned to you in case you want to update the testing instructions as I mentioned above before moving the PR to QA.

triple0t commented:
>  I'm approving this PR but leaving it assigned to you in case you want to update the testing instructions as I mentioned above before moving the PR to QA.

@rodrigoprimo, Done. 
Thank you

Aschepikov commented:
Hi @triple0t  I have some console error and I can't see coupon settings, already tried to clean the cache. Can you take a look pls and resolve conflicts too 
`Uncaught TypeError: Cannot read properties of undefined (reading 'toJSON')
    at o.templateContext (newsletter_editor.js?ver=4.5.0:2:244203)
    at Function.result (newsletter_editor.js?ver=4.5.0:2:1384816)
    at o.mixinTemplateContext (newsletter_editor.js?ver=4.5.0:2:267053)
    at o._renderTemplate (newsletter_editor.js?ver=4.5.0:2:266817)
    at o.render (newsletter_editor.js?ver=4.5.0:2:266642)
    at Je (newsletter_editor.js?ver=4.5.0:2:1376475)
    at o.<anonymous> (newsletter_editor.js?ver=4.5.0:2:1376903)
    at o.render (newsletter_editor.js?ver=4.5.0:2:1370041)
    at o.changeSettings (newsletter_editor.js?ver=4.5.0:2:135581)
    at v (newsletter_editor.js?ver=4.5.0:2:290365)`
![image](https://user-images.githubusercontent.com/113185437/215430985-a0b6d774-9d36-4e59-a831-fe1544561214.png)

Aschepikov reviewed (approved):
GJ!
